Scheduler extension

ravigadde opened the pull request:
Implementation of scheduler extension proposal as discussed in #11470

Commit:
Scheduler extension

ravigadde commented:
cc @davidopp @bgrant0607 

Please review when you get a chance.

ravigadde commented:
@davidopp Any feedback? Can you please review after 1.1 freeze?

davidopp commented:
I briefly took a peek at this last night and am planning to review this by end of tomorrow. Very sorry for the delay.

davidopp commented:
Very sorry for the delay. I will review it this Friday.

karlkfi commented:
I would like to see some kind of documentation included in this PR that describes the capability being added and how people might make use of it to externalize/extend scheduling logic.

davidopp commented:
+1

davidopp commented on plugin/pkg/scheduler/api/types.go:
```diff
@@ -26,6 +29,8 @@ type Policy struct {
 	// Holds the information to configure the priority functions
 	Priorities []PriorityPolicy `json:"priorities"`
+	// Holds the information to communicate with the extender
+	ExtenderConfig *ExtenderConfig `json:"extender"`
```
Should we allow an array of these, so the user can configure more than one?

ravigadde replied:
Make sense, will do.

davidopp commented on plugin/pkg/scheduler/api/types.go:
```diff
@@ -102,3 +107,44 @@ type LabelPreference struct {
+
+// Holds the parameters used to communicate with the extender. If a verb is unspecified/empty,
+// it is assumed that the extender chose not to provide that extension.
+type ExtenderConfig struct {
```
Please add a comment to each field of this struct.

ravigadde replied:
Ack

davidopp commented on plugin/pkg/scheduler/generic_scheduler.go:
```diff
@@ -166,20 +174,35 @@ func PrioritizeNodes(pod *api.Pod, podLister algorithm.PodLister, priorityConfig
+			return schedulerapi.HostPriorityList{}, err
+		}
+
+		prioritizedList, err := extender.Prioritize(pod, &nodes)
```
Is there some reason why you made this interface different from priorityFunc() ?
In particular, priorityFunc() has access to all pods (whereas extender.Prioritize() takes just the pod you are trying to schedule), and only the set of nodes that passed the predicate checks (whereas extender.Prioritize() takes the set of all nodes).

ravigadde replied:
I left out the pods intentionally from the API, because I felt any inter pod dependencies that affect scheduling should reside in the kube scheduler (ex: Anti affinity scheduling). Any extender that would like to have the pods can get it from apiserver too. But its open for debate, let me know what your preference is.

Regarding the set of nodes passed to extender, its the same set as priorityFunc.  PrioritizeNodes gets a fake minion lister with the filtered nodes.

davidopp replied:
Regarding first comment: I think the APIs should be the same. So I would suggest passing all the pods to the extender.

Regarding second comment: Sorry, you're right, never mind my comment.

davidopp commented:
Sorry it took me so long to take a look at this.

This looks OK to me, with all of the caveats @bgrant0607 mentioned here
https://github.com/kubernetes/kubernetes/issues/11470#issuecomment-134866698

That is -- you are going to need to make significant changes to this when we start making planned scheduler changes that we just haven't had time for yet (e.g. unified extension mechanism, and changing the way we choose the "best" machine from summing up priority scores to using a decision tree). Since we also want to make improvements in scheduler performance and generally make the scheduler implementation a bit cleaner, probably we will just create a new scheduler for all of this, and allow people to continue using the old scheduler but won't do new work on it.

Of course, before submitting you need to write unit tests and either e2e or integration test.

ravigadde commented:
@davidopp 
Thanks for the review. I am working on the feedback. Wrt test case comment, is it adequate if the test cases launch a http server which supports the extensions and shut it down after they are done?

@karlkfi 
I will add some documentation as well when I update the PR.

ravigadde commented:
@davidopp Addressed review comments, PTAL. I have implemented unit test only at this point. Need to spend more time on figuring out e2e/integ test cases for this.

davidopp commented on docs/design/scheduler_extender.md:
```diff
@@ -0,0 +1,100 @@
+
+# Scheduler extender
+
+Kubernetes scheduler schedules based on resources managed by Kubernetes. Scheduler extender is a way to extend the scheduling capability of k8s for resources that are not managed by k8s. A couple of example use cases are storage affinity and network availability.
```
I think this needs a bit more introduction, and actually shouldn't mention "resources that are not managed by k8s" because there are other ways to handle that (e.g. write your own scheduler). I'd just replace this paragraph with:

There are three ways to add new scheduling rules (predicates and priority functions) to Kubernetes: (1) by adding these rules to the scheduler and recompiling (described here: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/scheduler.md), (2) implementing your own scheduler process that runs instead of, or alongside of, the standard Kubernetes scheduler, (3) implementing a "scheduler extender" process that the standard Kubernetes scheduler calls out to as a final pass when making scheduling decisions. This document describes the third approach. (Note that the three approaches are not mutually exclusive.)

davidopp commented on docs/design/scheduler_extender.md:
```diff
@@ -0,0 +1,100 @@
+
+Kubernetes scheduler schedules based on resources managed by Kubernetes. Scheduler extender is a way to extend the scheduling capability of k8s for resources that are not managed by k8s. A couple of example use cases are storage affinity and network availability.
+
+When scheduling a pod, the extender allows an external process to filter and prioritize nodes. Two separate http/https calls are issued through the extender for "filter" and "prioritize" actions. Extender's configuration is part of the scheduler policy file. The configuration specifies how to reach the extender and whether to use http or https.
```
s/issued through the extender for "filter" and "prioritize" actions/issued to the extender, one for "filter" and one for "prioritize."/

s/Extender's configuration...file/To use the extender, you must create a scheduler policy configuration file./

davidopp commented on docs/design/scheduler_extender.md:
```diff
@@ -0,0 +1,100 @@
+type ExtenderConfig struct {
+	// URLPrefix at which the extender is available
+	URLPrefix string `json:"urlPrefix"`
+	// Verb for the filter call, empty if not supported
```
This needs more explanation. What does the string I put here do?

davidopp commented on docs/design/scheduler_extender.md:
```diff
@@ -0,0 +1,100 @@
+	URLPrefix string `json:"urlPrefix"`
+	// Verb for the filter call, empty if not supported
+	FilterVerb string `json:"filterVerb,omitempty"`
+	// Verb for the prioritize call, empty if not supported
```
This needs more explanation. What does the string I put here do?

davidopp commented on docs/design/scheduler_extender.md:
```diff
@@ -0,0 +1,100 @@
+	EnableHttps bool `json:"enableHttps,omitempty"`
+	// TLSConfig specifies the transport layer security config
+	TLSConfig *client.TLSClientConfig `json:"tlsConfig,omitempty"`
+	// HTTPTimeout specifies the timeout duration for a call to the extender
```
Explain what is the scheduler behavior if the filter/prioritize call times out.

davidopp replied:
ping -- please do this.

ravigadde replied:
Done

davidopp commented on docs/design/scheduler_extender.md:
````diff
@@ -0,0 +1,100 @@
+}
+```
+
+The "filter" call may further prune the set of nodes based on its predicates. Scores returned by the "prioritize" call are added to the k8s scores (computed through its priority functions) and used for final host selection.
````
s/further prune/prune/

Also, add the struct that represents what is returned.

davidopp commented on plugin/pkg/scheduler/generic_scheduler.go:
```diff
@@ -166,20 +176,37 @@ func PrioritizeNodes(pod *api.Pod, podLister algorithm.PodLister, priorityConfig
+		}
+
+		for _, extender := range extenders {
+			prioritizedList, err := extender.Prioritize(pod, &nodes)
```
priorityFunc() (L177) gets passed a podLister which allows the priority function to list all pods. To make the extender priority function interface as similar to the regular priority function interface as possible, I think you should pass all pods here, not just the pod being scheduled.

ravigadde replied:
@davidopp Thanks for the review.

The concern here is that the pod list can be too large. The other assumption I made is that any inter pod synergies for scheduling purpose can be handled by the k8s scheduler (affinity, anti affinity etc).

davidopp commented on plugin/pkg/scheduler/generic_scheduler.go:
```diff
@@ -166,20 +176,37 @@ func PrioritizeNodes(pod *api.Pod, podLister algorithm.PodLister, priorityConfig
+		for _, extender := range extenders {
+			prioritizedList, err := extender.Prioritize(pod, &nodes)
+			if err != nil {
+				return schedulerapi.HostPriorityList{}, err
```
I think in this case I think you should just "continue". Prioritization is by definition best-effort, so skipping priority extenders that generate an error seems fine. At least you will be able to use the built-in prioritization.

davidopp commented on plugin/pkg/scheduler/extender_test.go:
```diff
@@ -0,0 +1,229 @@
+/*
```
Please have at least one test that has multiple (two is fine) extenders.

ravigadde replied:
Done

davidopp commented on plugin/pkg/scheduler/generic_scheduler.go:
```diff
@@ -166,20 +176,37 @@ func PrioritizeNodes(pod *api.Pod, podLister algorithm.PodLister, priorityConfig
+			}
+
+			for _, hostEntry := range *prioritizedList {
+				combinedScores[hostEntry.Host] += hostEntry.Score
```
I think you should add weight as one of the config options in the config file.

davidopp replied:
(and then use it here)

davidopp commented on plugin/pkg/scheduler/generic_scheduler.go:
```diff
@@ -138,6 +139,15 @@ func findNodesThatFit(pod *api.Pod, podLister algorithm.PodLister, predicateFunc
+			if err != nil {
+				return api.NodeList{}, FailedPredicateMap{}, err
+			}
+			filtered = filteredList.Items
```
after this line, maybe
if len(filtered) == 0 break

ravigadde replied:
Done

davidopp commented on plugin/pkg/scheduler/generic_scheduler.go:
```diff
@@ -138,6 +139,15 @@ func findNodesThatFit(pod *api.Pod, podLister algorithm.PodLister, predicateFunc
 	}
+	if len(filtered) > 0 && len(extenders) != 0 {
+		for _, extender := range extenders {
+			filteredList, err := extender.Filter(pod, &api.NodeList{Items: filtered})
```
should you also pass the full machinetoPods map, since predicate() on L121 gets it (for the node on the current iteration)?

ravigadde replied:
Same concern as above with sending a large pod list

davidopp replied:
Yeah, this is a really painful decision. Without passing the list of pods running on each machine, the extender writer is really limited in what they can do. For example, they can't do anything that requires knowing the free and used resources on the machine (since that requires knowing the requests of the pods on the machine), or implement a policy that avoids conflicts between the pod-to-schedule and already-running pods. OTOH I think the whole mechanism will be impossibly slow if we pass the list of pods on each machine. This is why this mechanism is really a bad way to extend the scheduler. I wish you would consider some other approach, like figuring out a way we can dynamically link scheduling policies (predicates and priorities) into the existing scheduler. (Dynamically link meaning requires restarting the scheduler process but not recompiling.) Since we already agreed to let you do this, I'm going to say what you have here is OK, but I have a feeling you're the only person who is ever going to use this extension mechanism.

eparis replied:
I feel the opposite, when you give someone such a simple interface they are going to use it, rather than 'doing it right'...

ravigadde replied:
I would like to have a clean split of responsibilities between k8s scheduler and the extender. k8s will do the cpu/mem/label/... based scheduling. extender is limited to solving use cases beyond what k8s provides, like storage and network aware scheduling. I dont want the extender to be very rich, then it gravitates towards being another scheduler on its own.

timothysc replied:
> like storage and network aware scheduling. I dont want the extender to be very rich, then it gravitates towards being another scheduler on its own.

that statement is antithetical.

ravigadde replied:
As I was explaining in other comments, the intention is to have a split of responsibilities, not overlapping.

davidopp commented:
Sorry for the delay in re-reviewing. If the difficulty in writing an e2e test is that it's hard to add a new "extender" component maybe an integration test wold be easier to add a new component to; see test/integration/scheduler_test.go for an example

ravigadde commented:
@davidopp Updated the review with feedback comments and integ test case. PTAL when you get a chance.

davidopp commented on plugin/pkg/scheduler/api/types.go:
```diff
@@ -100,3 +108,59 @@ type LabelPreference struct {
+	EnableHttps bool `json:"enableHttps,omitempty"`
+	// TLSConfig specifies the transport layer security config
+	TLSConfig *client.TLSClientConfig `json:"tlsConfig,omitempty"`
+	// HTTPTimeout specifies the timeout duration for a call to the extender
```
Explain what is the scheduler behavior if the filter/prioritize call times out.

ravigadde replied:
Done

davidopp commented on plugin/pkg/scheduler/api/v1/types.go:
```diff
@@ -100,3 +108,50 @@ type LabelPreference struct {
+	// TLSConfig specifies the transport layer security config
+	TLSConfig *client.TLSClientConfig `json:"tlsConfig,omitempty"`
+	// HTTPTimeout specifies the timeout duration for a call to the extender
+	HTTPTimeout time.Duration `json:"httpTimeout,omitempty"`
```
Explain what is the scheduler behavior if the filter/prioritize call times out.

ravigadde replied:
Done

davidopp commented on docs/design/scheduler_extender.md:
```diff
@@ -0,0 +1,116 @@
+
+When scheduling a pod, the extender allows an external process to filter and prioritize nodes. Two separate http/https calls are issued to the extender, one for "filter" and one for "prioritize" actions. To use the extender, you must create a scheduler policy configuration file. The configuration specifies how to reach the extender, whether to use http or https and the timeout. Extender timeouts  ...
+
+Extender timeouts are treated
```
I think you meant to delete this sentence fragment.

ravigadde replied:
Oops.. done

davidopp commented on docs/design/scheduler_extender.md:
```diff
@@ -0,0 +1,116 @@
+
+There are three ways to add new scheduling rules (predicates and priority functions) to Kubernetes: (1) by adding these rules to the scheduler and recompiling (described here: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/scheduler.md), (2) implementing your own scheduler process that runs instead of, or alongside of, the standard Kubernetes scheduler, (3) implementing a "schedu ...
+
+When scheduling a pod, the extender allows an external process to filter and prioritize nodes. Two separate http/https calls are issued to the extender, one for "filter" and one for "prioritize" actions. To use the extender, you must create a scheduler policy configuration file. The configuration specifies how to reach the extender, whether to use http or https and the timeout. Extender timeouts  ...
```
This last sentence isn't completely correct. I think you can delete it now that you explain the correct behavior in the comment for HTTPTimeout.

timothysc replied:
I'm still flummoxed by the argument of why this shouldn't just be a separate scheduler.  https://github.com/kubernetes/kubernetes/pull/17197  proposal would not interfear with existing scheduler and therefore not violate any support on said cluster.

ravigadde replied:
@davidopp Done
@timothysc 
Sorry if this looks a little long, but here you go.

Its not about running one or many schedulers. Scheduling should be based on all resources in the cluster. The base k8s scheduler does a good job of scheduling based on cpu/mem/affinity etc. What is left out is storage and network resource based scheduling. When storage is remote (like GCEPersistentDisk, EBS etc), its not a scheduling constraint. But if you are on a hypercoverged infrastructure, it is absolutely. Placing workloads local/close to storage does wonders for performance. Similarly, network bandwidth is not an infinite resource on a server. When scheduling a pod (irrespective of which mechanism is used to achieve this), all of these are important requirements for performance sensitive applications.

Now lets talk about the mechanisms. As the document says, there are three mechanisms to do this. I will quote that verbtim here.
```
There are three ways to add new scheduling rules (predicates and priority functions) to Kubernetes: (1) by adding these rules to the scheduler and recompiling (described here: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/scheduler.md), (2) implementing your own scheduler process that runs instead of, or alongside of, the standard Kubernetes scheduler, (3) implementing a "scheduler extender" process that the standard Kubernetes scheduler calls out to as a final pass when making scheduling decisions. This document describes the third approach. (Note that the three approaches are not mutually exclusive.)
```
You are very familiar with Option 1. #17197 is way of accomplishing Option 2. So why is there a need for a third method? What I have described above doesn't fit in Options 1 or 2. In Option 1, k8s scheduler is not aware of all storage, network resources in the cluster. In Option 2, responsibility is split up at a pod level. What Option 3 provides is for the responsibility to be split up at a resource level. k8s schedules based on the resources it knows and the extenders schedule based on the resources they know and manage. Makes sense?

ravigadde replied:
k8s scheduler and the extenders work together to ensure all resources are accounted for when making scheduling decisions.

timothysc replied:
Totally familiar with the state space your in, but I'm against this approach.

If the primary motivation is to have arbitrary resources (labels), and apply your own arbitrary constraints (best fit), then imho there are much better ways we can approach this such that they are beneficial to the whole community.  Happy to do a hangout on the topic.

ravigadde replied:
labels have limitations, happy to do a hangout. Will connect on slack.

davidopp commented on docs/design/scheduler_extender.md:
````diff
@@ -0,0 +1,116 @@
+}
+```
+
+Arguments passed to the extender are the set of nodes filtered through the k8s predicates and the pod.
````
This is close but not exactly right. I believe it's:

Arguments passed to the FilterVerb endpoint on the extender are the set of nodes filtered through the k8s predicates and the pod. Arguments passed to the PrioritizeVerb endpoint on the extender are the set of nodes filtered through the k8s predicates and extender predicates and the pod.

ravigadde replied:
Done

davidopp commented on plugin/pkg/scheduler/algorithm/scheduler_interface.go:
```diff
@@ -18,10 +18,25 @@ package algorithm
+	// expected to be a subset of the supplied list.
+	Filter(pod *api.Pod, nodes *api.NodeList) (*api.NodeList, error)
+
+	// Prioritize based on extender implemented priority functions. Weight*priority is added
```
s/extender implemented/extender-implemented/

I'm not sure the sentence "Weight*priority is added up for each such priority function" makes sense where it is, and should probably be deleted.

ravigadde replied:
Done

davidopp commented on plugin/pkg/scheduler/algorithm/scheduler_interface.go:
```diff
@@ -18,10 +18,25 @@ package algorithm
+	Filter(pod *api.Pod, nodes *api.NodeList) (*api.NodeList, error)
+
+	// Prioritize based on extender implemented priority functions. Weight*priority is added
+	// up for each such priority function. The returned score & weight are used to compute
```
s/score & weight/scores & weight/

davidopp commented on plugin/pkg/scheduler/algorithm/scheduler_interface.go:
```diff
@@ -18,10 +18,25 @@ package algorithm
+
+	// Prioritize based on extender implemented priority functions. Weight*priority is added
+	// up for each such priority function. The returned score & weight are used to compute
+	// the weighted score for an extender. The weighted score is added to the score computed
```
s/The weighted score is added to the score/The weighted scores are added to the scores/

ravigadde replied:
Done

davidopp commented on plugin/pkg/scheduler/algorithm/scheduler_interface.go:
```diff
@@ -18,10 +18,25 @@ package algorithm
+	// Prioritize based on extender implemented priority functions. Weight*priority is added
+	// up for each such priority function. The returned score & weight are used to compute
+	// the weighted score for an extender. The weighted score is added to the score computed
+	// by Kubernetes scheduler. The total score is used to do the host selection.
```
s/The total score is used/The total scores are used/

ravigadde replied:
Done

davidopp commented on plugin/pkg/scheduler/api/types.go:
```diff
@@ -100,3 +108,61 @@ type LabelPreference struct {
+	Error string       `json:"error,omitempty"`
+}
+
+// HostPriority represents the priority of scheduling to a particular host, lower priority is better.
```
Actually higher priority is better.

ravigadde replied:
Yep, fixed it.

davidopp commented on plugin/pkg/scheduler/api/v1/types.go:
```diff
@@ -100,3 +108,52 @@ type LabelPreference struct {
+	Nodes apiv1.NodeList `json:"nodes"`
+}
+
+// HostPriority represents the priority of scheduling to a particular host, lower priority is better.
```
Actually higher priority is better.

davidopp commented:
Have you tested this out on a reasonably-sized cluster with a reasonable scheduling workload, to verify that the overhead of the callouts to the extender endpoint doesn't make the scheduler performance completely horrible? For example, considering that we don't do any caching in the scheduler right now, if there is one pending pod the scheduler will sit in a tight loop calling out to the extender over and over again.

BTW just to reiterate something we said in our offline discussion, we are making no guarantee that when we make future improvements to the scheduler (including, but not limited to, rewriting the scheduler), setups that use an extender will be able to benefit from them. We won't break the extender mechanism but we may not try to make it benefit from future changes either.

As I mentioned in a comment, I'd strongly prefer that you try to implement dynamic loading of scheduler policies into the scheduler (e.g. so that you can add priority and predicate functions with just a scheduler restart, without recompiling the scheduler) rather than the extension mechanism you are implementing here, which requires you to "pick your poison" (either send an untenably huge amount of information to the remote endpoint on every scheduler pass, or severely limit the types of priority and predicate functions you can write in the extension -- see my comment in the code for more explanation).

That said, we told you this would be OK so I'll accept this (once you make the changes in the comments I just added), but it does pain me to do it, and I suspect you're the only person who will ever use this feature.

eparis commented on plugin/pkg/scheduler/api/types.go:
```diff
@@ -100,3 +108,61 @@ type LabelPreference struct {
+	// fails the scheduling of the pod. Prioritize timeout is ignored, k8s/other extenders
+	// priorities are used to select the node.
+	HTTPTimeout time.Duration `json:"httpTimeout,omitempty"`
+}
```
Completely unrelated, but if we moved things like this to their own file we could have the docs be automatically generated/updated...

Not sure if we want to introduce that pattern into the project, as it hides the `type` away from the code author...   I wonder how hard it would be to get the munger to handle it in place...

No need to address this comment, just thinking outloud...

eparis commented on plugin/pkg/scheduler/api/v1/types.go:
```diff
@@ -100,3 +108,52 @@ type LabelPreference struct {
+	// fails the scheduling of the pod. Prioritize timeout is ignored, k8s/other extenders
+	// priorities are used to select the node.
+	HTTPTimeout time.Duration `json:"httpTimeout,omitempty"`
+}
```
Just starting to look, so ignore the dumb question as I keep looking, but why does `plugin/pkg/scheduler/api/types.go::ExtenderConfig` have a `Weight` but this doesn't? 

/me is still trying to figure out what the `Weight` means over there...

ravigadde replied:
Thanks, I missed it.

timothysc commented on docs/design/scheduler_extender.md:
```diff
@@ -0,0 +1,116 @@
+
+# Scheduler extender
+
+There are three ways to add new scheduling rules (predicates and priority functions) to Kubernetes: (1) by adding these rules to the scheduler and recompiling (described here: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/scheduler.md), (2) implementing your own scheduler process that runs instead of, or alongside of, the standard Kubernetes scheduler, (3) implementing a "schedu ...
```
If possible, please describe the phases and why it "extends" @ phase x.

ravigadde replied:
Good suggestion, done.

timothysc commented on plugin/pkg/scheduler/algorithm/scheduler_interface.go:
```diff
@@ -18,10 +18,25 @@ package algorithm
+	// up for each such priority function. The returned score & weight are used to compute
+	// the weighted score for an extender. The weighted score is added to the score computed
+	// by Kubernetes scheduler. The total score is used to do the host selection.
+	Prioritize(pod *api.Pod, nodes *api.NodeList) (*schedulerapi.HostPriorityList, int, error)
```
I debate splitting hairs here and calling this Rank vs. Prioritize..

ravigadde replied:
Good suggestion, @davidopp what do you think? Before I change names all over the place, want to make sure there is consensus.

davidopp replied:
I don't have a strong opinion, though "prioritize" seems slightly better since we call them "priority functions" so it sounds a little more consistent.

ashcrow commented on plugin/pkg/scheduler/extender_test.go:
```diff
@@ -0,0 +1,302 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
```
nit: It's 2015 :smile:

ravigadde replied:
Thanks, fixed them.

timothysc commented on plugin/pkg/scheduler/extender.go:
```diff
@@ -0,0 +1,175 @@
+}
+
+// Helper function to send messages to the extender
+func (h *HTTPExtender) send(action string, args interface{}, result interface{}) error {
```
Why does an extender need to send?  You're basically going to grind your scheduling to crawl.  As it stands today scheduling is already relatively slow due to several factors.

ravigadde replied:
Please see my other comments on need and the performance.

hongchaodeng replied:
Why not return "result" instead of putting it in arguments?

ravigadde replied:
Dont know the type. Its similar to Unmarshal.

ashcrow commented on plugin/pkg/scheduler/api/types.go:
```diff
@@ -100,3 +108,61 @@ type LabelPreference struct {
+
+type HostPriorityList []HostPriority
+
+func (h HostPriorityList) Len() int {
```
I realize that the previous iteration of these didn't have Godoc either but they really should as they are exported.

ravigadde replied:
Did you mean the package level comments in doc.go?

ravigadde replied:
Done

ashcrow commented on plugin/pkg/scheduler/api/v1/types.go:
```diff
@@ -100,3 +108,52 @@ type LabelPreference struct {
+
+type HostPriorityList []HostPriority
+
+func (h HostPriorityList) Len() int {
```
Same Godoc comment for these.

ravigadde replied:
Done

timothysc commented:
IMHO I'm in the camp that what you really want is your own scheduler here.  The idea is schlepping off curated results over a socket to a separate process which then performs more actions is kind of the antithesis of what most scheduling systems try to do.  They are purpose built hot-rods optimized for a workload/use case.  This appears like an attempt to add a generalization in order to allow for a specific specialization, data-locality - "(example: place a pod where its storage is)".  

Perhaps I'm way off the mark, but this seems like it would be an excellent topic for the scheduling SIG.  /cc @davidopp @karlkfi

ravigadde commented:
@davidopp 
Regarding your comment on performance. Yes this will do slower than the native call because of the overhead of making a callout. Dynamic loading doesn't solve this problem though, because the fundamental problem is that these resources are not known to k8s.

We are colocating the extender on the same node as the scheduler, performs well.

@timothysc 
Please refer to my earlier comment about a need for Option 3. It is not just the data locality use case, there are other use cases around networking, performance etc.

I will also say this one thing, perhaps TIC :). This is a platform for running microservices. Are we saying that scheduler should be monolithic and cant be run as a microservice?

timothysc commented:
@davidopp So I chatted with ravi to get a better understanding to what he's doing and why.  It "sounds like" a temporary workaround to enable features that I've longed for in the core for some time, but have had pretty much 0 time to implement.  http://people.cs.uchicago.edu/~chliu/doc/liuc_matchmaking.pdf 

Ideally we want composable constraint language to expressively match based on well defined semantics.  Predicates and labels in their current form simply won't cut it. 

My question then becomes: "What is our policy for deprecating temporary work-arounds"?

ravigadde commented:
@timothysc 
Thanks for taking the time to jump on a call, really appreciate it. I have some disagreement on the term "temporary workaround" though. There are two types of use cases, atleast for us:
1) We need XYZ, k8s scheduler doesn't do it today. Lets implement it in the extender while community rallies around common ground for XYZ. The usage of extender is temporary in this case.
2) We need this Super complicated thing that is so specific to our platform and has no chance of ever making it to k8s. For this kind of use case, the usage of extender is of a permanent nature.

ravigadde commented:
No system solves 100% of use cases, having a pluggable system makes it easier for greater adoption. This is "a" way to make the scheduler pluggable. plugins by nature do small, specific things tailored to that platform/use case.

davidopp commented:
@timothysc : @bgrant0607 and I had a long discussion with @ravigadde about this approach several months ago, before he started working on it. He convinced us that there really isn't any alternative given their requirements. We felt that within certain constraints, it is OK to accept this. The certain constraints are outlined in these comments:
https://github.com/kubernetes/kubernetes/pull/13580#issuecomment-143598386
https://github.com/kubernetes/kubernetes/issues/11470#issuecomment-134866698
https://github.com/kubernetes/kubernetes/pull/13580#issuecomment-156395598 (second paragraph)

davidopp commented:
> BTW just to reiterate something we said in our offline discussion, we are making no guarantee that when we make future improvements to the scheduler (including, but not limited to, rewriting the scheduler), setups that use an extender will be able to benefit from them. We won't break the extender mechanism but we may not try to make it benefit from future changes either.

Actually, to clarify this further: When we rewrite the scheduler, it may not support the extender mechanism. When I said "we will not break the extender mechanism" I meant when we make changes to the existing scheduler.

davidopp commented:
(That said, there is no immediate plan to rewrite the scheduler, as nobody has time.)

pmorie commented on contrib/mesos/pkg/scheduler/plugin.go:
```diff
@@ -250,7 +250,7 @@ func recoverAssignedSlave(pod *api.Pod) string {
 // Schedule implements the Scheduler interface of Kubernetes.
 // It returns the selectedMachine's name and error (if there's any).
-func (k *kubeScheduler) Schedule(pod *api.Pod, unused algorithm.NodeLister) (string, error) {
+func (k *kubeScheduler) Schedule(pod *api.Pod, unused algorithm.NodeLister, extenders []algorithm.SchedulerExtender) (string, error) {
```
Why not make `unused` a don't-care:

```go
func (k *kubeScheduler) Schedule(pod *api.Pod, _ algorithm.NodeLister, extenders []algorithm.SchedulerExtender) (string, error)
```

ravigadde replied:
This one existed previously.

pmorie commented on docs/design/scheduler_extender.md:
```diff
@@ -0,0 +1,115 @@
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Scheduler extender
```
Is there a reason not to call this an 'extension' ?  We use nouns for things like this elsewhere in the system (viz: 'plugin').

davidopp replied:
Although I'm sure it wasn't the original intention, one reason to call it an "extender" might be so that we can reserve the term "extension" for the unified extension architecture we will hopefully develop eventually (to cover all call-out extensions like this one, admission controller extension, etc.).

pmorie commented on plugin/pkg/scheduler/algorithm/scheduler_interface.go:
```diff
@@ -18,10 +18,24 @@ package algorithm
+	schedulerapi "k8s.io/kubernetes/plugin/pkg/scheduler/api"
 )
 
+// SchedulerExtender is an abstract interface for external processes to influence scheduling
```
nit: s/abstract//, interfaces are inherently abstract

pmorie replied:
It would be helpful to mention the use-case of third-party resources influencing scheduling

ravigadde replied:
Done.

pmorie commented on plugin/pkg/scheduler/algorithm/scheduler_interface.go:
```diff
@@ -18,10 +18,24 @@ package algorithm
+	Prioritize(pod *api.Pod, nodes *api.NodeList) (*schedulerapi.HostPriorityList, int, error)
+}
+
 // Scheduler is an interface implemented by things that know how to schedule pods
```
While you're in here you might as well fix this comment to refer to the right type name (ScheduleAlgorith)

ravigadde replied:
Done

pmorie commented on plugin/pkg/scheduler/api/v1/types.go:
```diff
@@ -100,3 +108,54 @@ type LabelPreference struct {
+	// Verb for the filter call, empty if not supported. This verb is appended to the URLPrefix when issuing the filter call to extender.
+	FilterVerb string `json:"filterVerb,omitempty"`
+	// Verb for the prioritize call, empty if not supported. This verb is appended to the URLPrefix when issuing the prioritize call to extender.
+	PrioritizeVerb string `json:"prioritizeVerb,omitempty"`
```
It would be helpful to know the precedence of `FilterVerb` and `PrioritizeVerb` relative to one another

ravigadde replied:
There is no precedence, they work very similar to the base scheduler. Filter does hard constraint matching, prioritize assigns (weighted) scores to the remaining nodes. If the config has a non empty string, it is assumed that the extender is interested in that call out. If not, the call out is skipped.

pmorie commented on plugin/pkg/scheduler/api/validation/validation.go:
```diff
@@ -34,5 +34,10 @@ func ValidatePolicy(policy schedulerapi.Policy) error {
 		}
 	}
 
+	for _, extender := range policy.ExtenderConfigs {
```
Is it an error if you enable https but have a nil TLS config?

liggitt replied:
not positive on the context, but presumably a nil TLSConfig means use the defaults (system roots, hostname verification on)

ravigadde replied:
Insecure mode is set which skips verification. Look at TLSConfigFor in client package.

pmorie commented on plugin/pkg/scheduler/api/v1/types.go:
```diff
@@ -100,3 +108,54 @@ type LabelPreference struct {
+// HostPriority represents the priority of scheduling to a particular host, higher priority is better.
+type HostPriority struct {
+	Host  string `json:"host"`
+	Score int    `json:"score"`
```
Godoc for `ExtenderArgs` and `HostPriority` fields for completeness :)

ravigadde replied:
Done

pmorie commented on plugin/pkg/scheduler/extender.go:
```diff
@@ -0,0 +1,175 @@
+		return &http.Transport{
+			TLSClientConfig: tlsConfig,
+		}, nil
+	} else {
```
Nit: else block not needed

ravigadde replied:
Done

pmorie commented on plugin/pkg/scheduler/extender.go:
```diff
@@ -0,0 +1,175 @@
+
+	if err := h.send(h.filterVerb, &args, &result); err != nil {
+		return nil, err
+	} else {
```
Else block isn't needed here.

ravigadde replied:
Done

pmorie commented on plugin/pkg/scheduler/extender.go:
```diff
@@ -0,0 +1,175 @@
+
+	if err := h.send(h.prioritizeVerb, &args, &result); err != nil {
+		return nil, 0, err
+	} else {
```
else block isn't needed here

ravigadde replied:
Done

davidopp commented:
>  Are we saying that scheduler should be monolithic and cant be run as a microservice?

I think "can't" would be too strong a word. It's a question of tradeoffs. I think long-term we will encourage three ways for people to extend the scheduler
(1) compile it directly into the upstream scheduler; select desired policies at runtime via config file
(2) implement their own scheduler, either for all pods or a subset of pods. to aid in this, we will create libraries out of the default/standard kubernetes scheduler's priority and predicate functions, so that someone who writes a new scheduler can easily create the same kind of "all the standard logic plus the extension/addon logic" behavior you get with the approach you're using in this PR
(3) a variant of (1) where we find a way people can dynamically link their policies into the upstream scheduler

I guess there is also a fourth approach where we make the upstream scheduler's policy language so generic, and able to use third party resources, that every possible policy can be expressed via the scheduler config file and doesn't require any actual coding (this seems along the lines of what Tim was suggesting with ClassAds, if I understood correctly).

All of these are going to have better performance than the microservice approach: (1)/(3)/(4) have no network communication, while (2) communicates much less information over the network in the general case than the callout approach (scheduler-to-master is just a binding, and master-to-scheduler is the full cluster state but delivered via watch rather than having to send all the information for every pod and node ever time you try to schedule a pod). Also, (1)/(3)/(4) allow for optimizations that are harder in the callout approach, such as caching results (difficult in the callout approach because the main scheduler doesn't know when the remote process may have been restarted with new logic).

Anyway, I think we should move the philosophical discussion elsewhere, so as to speed up the remaining review for this PR.

hongchaodeng commented on docs/design/scheduler_extender.md:
````diff
@@ -0,0 +1,115 @@
+
+When scheduling a pod, the extender allows an external process to filter and prioritize nodes. Two separate http/https calls are issued to the extender, one for "filter" and one for "prioritize" actions. To use the extender, you must create a scheduler policy configuration file. The configuration specifies how to reach the extender, whether to use http or https and the timeout.
+
+```
````
use \`\`\`Go for syntax highlight?

ravigadde replied:
Done

hongchaodeng commented on docs/design/scheduler_extender.md:
````diff
@@ -0,0 +1,115 @@
+
+A sample scheduler policy file with extender configuration:
+
+```
````
use \`\`\`json for syntax highlight?

ravigadde replied:
Done

hongchaodeng commented on docs/design/scheduler_extender.md:
````diff
@@ -0,0 +1,115 @@
+
+Arguments passed to the FilterVerb endpoint on the extender are the set of nodes filtered through the k8s predicates and the pod. Arguments passed to the PrioritizeVerb endpoint on the extender are the set of nodes filtered through the k8s predicates and extender predicates and the pod.
+
+```
````
use \`\`\`Go?

ravigadde replied:
Done

hongchaodeng commented on contrib/mesos/pkg/scheduler/plugin.go:
```diff
@@ -250,7 +250,7 @@ func recoverAssignedSlave(pod *api.Pod) string {
 
```
This [commit](https://github.com/kubernetes/kubernetes/commit/2c4142494a4e2d61f36814341a38f9b1a69745de) changed "plugin.go" to "podreconciler.go".

It would require a rebase.

hongchaodeng commented on plugin/pkg/scheduler/algorithm/types.go:
```diff
@@ -18,35 +18,13 @@ package algorithm
 // FitPredicate is a function that indicates if a pod fits into an existing node.
 type FitPredicate func(pod *api.Pod, existingPods []*api.Pod, node string) (bool, error)
 
-// HostPriority represents the priority of scheduling to a particular host, lower priority is better.
```
A lot of diffs are caused by the move-around of these API types.
Can we separate these changes to another PR? This can help make the review clearer here :)

ravigadde replied:
Too much to do that now. We already went through a few rounds of reviews.

hongchaodeng commented on docs/design/scheduler_extender.md:
```diff
@@ -0,0 +1,115 @@
+
+# Scheduler extender
+
+There are three ways to add new scheduling rules (predicates and priority functions) to Kubernetes: (1) by adding these rules to the scheduler and recompiling (described here: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/scheduler.md), (2) implementing your own scheduler process that runs instead of, or alongside of, the standard Kubernetes scheduler, (3) implementing a "schedu ...
```
AFAICT, we can improve the docs by separating it into two sub-sections
- Motivation
- Technical details/approaches. This part could be further separated into one about algorithm/workflow and another talking about API changes?

hongchaodeng replied:
Is it common to lay out goals and non-goals in k8s docs? Another nice thing to talk about is that lay out goals such as:
- Allow external process to access generic scheduler API, including filtering and prioritization.

ravigadde replied:
Thanks for the feedback, will address this separately in another PR.

hongchaodeng commented on docs/design/scheduler_extender.md:
```diff
@@ -0,0 +1,115 @@
+	// URLPrefix at which the extender is available
+	URLPrefix string `json:"urlPrefix"`
+	// Verb for the filter call, empty if not supported. This verb is appended to the URLPrefix when issuing the filter call to extender.
+	FilterVerb string `json:"filterVerb,omitempty"`
```
How about "PredicateVerb"? This is to make naming consistent

hongchaodeng commented on docs/design/scheduler_extender.md:
```diff
@@ -0,0 +1,115 @@
+	PrioritizeVerb string `json:"prioritizeVerb,omitempty"`
+	// The numeric multiplier for the node scores that the prioritize call generates.
+	// The weight should be a positive integer
+	Weight int `json:"weight,omitempty"`
```
There seems to be only one value of weight. But in the following sample policy I saw many prioritity functions. How do they correlate?

davidopp replied:
Each extender has one priority function and one predicate function.

hongchaodeng replied:
What prevents each extender to have multiple filter and priority functions? It's one HTTP roundtrip :)

hongchaodeng replied:
So a way to model this is that one extender is an isolated identity that provides some extension functions to do the work.

davidopp replied:
> What prevents each extender to have multiple filter and priority functions? It's one HTTP roundtrip :)

I think it would just make things more complicated. I think it's simpler if there's one prioritize endpoint that runs one prioritize function, and one filter endpoint that runs one predicate function. Since they are presumably in the same process, it should not be hard to combine the logic into one predicate function and one priority function, so I don't think this overly restricts the generality of the extender.

hongchaodeng commented on docs/design/scheduler_extender.md:
````diff
@@ -0,0 +1,115 @@
+```
+// Holds the parameters used to communicate with the extender. If a verb is unspecified/empty,
+// it is assumed that the extender chose not to provide that extension.
+type ExtenderConfig struct {
````
How does this config correlates to the configuration file? It would be easier for users to understand if there would be some examples. Can you add it?

davidopp replied:
Yeah, extending
https://github.com/kubernetes/kubernetes/blob/master/examples/scheduler-policy-config.json
(or creating a separate example file called scheduler-policy-config-with-extender.json) to include an example of using an extender might be good.

ravigadde replied:
Done

hongchaodeng commented on docs/design/scheduler_extender.md:
````diff
@@ -0,0 +1,115 @@
+}
+```
+
+Arguments passed to the FilterVerb endpoint on the extender are the set of nodes filtered through the k8s predicates and the pod. Arguments passed to the PrioritizeVerb endpoint on the extender are the set of nodes filtered through the k8s predicates and extender predicates and the pod.
````
We can improve the docs by adding two more sections here -- one for predicate and one for prioritization.

ravigadde replied:
Will take up in another PR

hongchaodeng commented on plugin/pkg/scheduler/algorithm/scheduler_interface.go:
```diff
@@ -18,10 +18,24 @@ package algorithm
+	// Prioritize based on extender-implemented priority functions. The returned scores & weight
+	// are used to compute the weighted score for an extender. The weighted scores are added to
+	// the scores computed  by Kubernetes scheduler. The total scores are used to do the host selection.
+	Prioritize(pod *api.Pod, nodes *api.NodeList) (*schedulerapi.HostPriorityList, int, error)
```
What's the second returned thing? Add names to returned arguments?

ravigadde replied:
Good suggestion, done.

hongchaodeng commented on plugin/pkg/scheduler/algorithm/scheduler_interface.go:
```diff
@@ -18,10 +18,24 @@ package algorithm
 // onto machines.
 type ScheduleAlgorithm interface {
-	Schedule(*api.Pod, NodeLister) (selectedMachine string, err error)
+	Schedule(*api.Pod, NodeLister, []SchedulerExtender) (selectedMachine string, err error)
```
This is assuming all schedule algorithms takes extensions, which is quite general. But I doubt that not all schedule algorithms has the same extension interface like filter and prioritize.

davidopp replied:
> This is assuming all schedule algorithms takes extensions

No; it can be empty.

> not all schedule algorithms has the same extension interface like filter and prioritize

I think this extension mechanism is just intended to work with schedulers that use scheduler.genericScheduler, hence they do use predicate and priority functions. I think this extension approach is reasonable given how things work today.

hongchaodeng replied:
Hi @davidopp . Thanks for explaining it. I'm still confused that for #11793 and #17197, if multi-schedulers is supported, wouldn't this interface add more burden to other schedulers?

I think filtering and prioritization are still pretty generic. So this is still good in general.

davidopp replied:
Oh, I see your point. I think I agree -- it would be better to make []SchedulerExtender be a member of struct genericScheduler
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/generic_scheduler.go#L55
rather than in the ScheduleAlgorithm interface, which is supposed to be more generic. I think this change would not be hard.

hongchaodeng replied:
Thanks for your understanding :)

ravigadde replied:
Ok

hongchaodeng commented on plugin/pkg/scheduler/extender.go:
```diff
@@ -0,0 +1,175 @@
+
+// Helper function to send messages to the extender
+func (h *HTTPExtender) send(action string, args interface{}, result interface{}) error {
+	out, err := json.Marshal(args)
```
json marshalling shouldn't be laid inside the send function? send() is like networking abstraction. Marshal things outside send() and pass bytes in?

ravigadde replied:
I think of it as a utility function to deal with sending the message and unmarshaling the response. If I rename the function, is it helpful. Dont want to repeat Unmarshal code.

hongchaodeng commented on plugin/pkg/scheduler/extender_test.go:
```diff
@@ -0,0 +1,302 @@
+	schedulerapi "k8s.io/kubernetes/plugin/pkg/scheduler/api"
+)
+
+type fitPredicate func(pod *api.Pod, node *api.Node) (bool, error)
```
This seems pretty generic.
Put it into pkg?

ravigadde replied:
Its specific to the extender. Others care about other pods running on the node. Extender only looks at the pod being scheduled and the nodes.

hongchaodeng commented on plugin/pkg/scheduler/extender_test.go:
```diff
@@ -0,0 +1,302 @@
+	return result, nil
+}
+
+type FakeExtender struct {
```
Always put structs on top of file.
Maybe create a new file for fake extender?

ravigadde replied:
I dont know if there is a formal convention. I see other instances where a structure and their methods are together (scheduler_test.go)

hongchaodeng commented on plugin/pkg/scheduler/extender_test.go:
```diff
@@ -0,0 +1,302 @@
+	return &result, f.weight, nil
+}
+
+func TestGenericSchedulerWithExtenders(t *testing.T) {
```
It's best practice in Go to put testing functions on top and other helper functions (e.g. errorPredicateExtender, etc.) in the bottom of the file.

ravigadde replied:
Same as above.

hongchaodeng commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -153,7 +153,20 @@ func (f *ConfigFactory) CreateFromConfig(policy schedulerapi.Policy) (*scheduler
 
-	return f.CreateFromKeys(predicateKeys, priorityKeys)
+	config, err := f.CreateFromKeys(predicateKeys, priorityKeys)
+	if err == nil && len(policy.ExtenderConfigs) != 0 {
```
```Go
if err != nil {
  return nil, err
}
```
Then
```Go

		for _, extender := range policy.ExtenderConfigs {
			glog.V(2).Infof("Creating extender with config %+v", extender)
			extender, err := scheduler.NewHTTPExtender(&extender, policy.APIVersion)
			if err != nil {
				return nil, err
			}
			config.Extenders = append(config.Extenders, extender)
		}
```

ravigadde replied:
This code will change because of the other comment about the interface.

ravigadde commented:
@davidopp 

Addressed all review comments and rebased to latest master. PTAL when you get a chance.

davidopp commented on plugin/pkg/scheduler/extender.go:
```diff
@@ -0,0 +1,172 @@
+func (h *HTTPExtender) send(action string, args interface{}, result interface{}) error {
+	out, err := json.Marshal(args)
+	if err != nil {
+		return nil
```
should this be "return err" ?

davidopp commented:
Looks good (I see you fixed my one comment just now). Squash the commits and then I'll LGTM.

ravigadde commented:
@davidopp Thanks a lot, just did.

davidopp commented:
ok to test

davidopp commented:
LGTM

davidopp commented:
@ravigadde integration test is failing with

```
01:18:56 # k8s.io/kubernetes/test/integration
01:18:56 _output/local/go/src/k8s.io/kubernetes/test/integration/extender_test.go:228: unknown "k8s.io/kubernetes/pkg/client/unversioned".Config field 'Version' in struct literal
01:18:56 FAIL	k8s.io/kubernetes/test/integration [build failed]
```

Did you try running the integration tests locally? `hack/test-integration.sh` should do it.

ravigadde commented:
@davidopp Let me update to latest and fix it. Looks like this commit renamed the field recently - https://github.com/kubernetes/kubernetes/commit/94ad6aa0c4651a584c8cf04aaf0866afa1f56f49

ravigadde commented:
@davidopp Done, all tests pass now

davidopp commented:
LGTM

davidopp commented:
Yay! Thanks for your patience with this, @ravigadde

ravigadde commented:
@davidopp Thank you for your effort to make this happen

khoi0107 commented:
@ravigadde Hi, is there sample implementation of an extender somewhere?

ravigadde commented:
@khoi0107 I haven't seen an open implementation. @ping035627 submitted a patch sometime ago, not sure if their implementation is open. If you have any specific questions I can help with, ping me on slack - ravi.gadde
